base/intfuncs.jl

```diff
@@ -488,7 +488,8 @@ function nextpow(a::Real, x::Real)
     n = ceil(Integer,log(a, x))
     # round-off error of log can go either direction, so need some checks
     p = a^(n-1)
-    x > typemax(p) && throw(DomainError(x,"argument is beyond the range of type of the base"))
+    hastypemax(typeof(p)) && x > typemax(p) &&
+        throw(DomainError(x,"argument is beyond the range of type of the base"))
```

Comment:
If we guarantee the error (having automatic type widening in Base is a non-starter, as far as I'm aware) this throwing behavior should be documented & tested.

Reply:
Isn't that a concern that's completely separate from this PR? It's not like this commit adds the `throw`, or makes it happen where it didn't before. I just added the `hastypemax(typeof(p))` check (with the goal of preventing spurious errors), so I don't really see how does the `throw` come into the picture here.

Reply:
If we're already going to merge a change for this code, why not take the opportunity to improve its documentation as well?

Reply:
OK.

Regarding the documentation change (guaranteeing the exception), that seems to depend on other changes (#49670 is related), because currently base two is a special case regarding output width:
```julia-repl
julia> prevpow(Int8(2), 800)
512

julia> typeof(ans)
Int64

julia> prevpow(Int8(3), 800)
ERROR: DomainError with 800:
argument is beyond the range of type of the base
```

I'll add some tests for base three, though.

Reply:
> If the `2` case is expected to stay special cased

Ideally it wouldn't be special-cased, at least not to the degree that we mention that in the docs, IMO.

Reply:
Agreed, but I was talking about the behavior that `nextpow`/`prevpow` etc. throw if the result is not representable in the input type :) With #49672, the special casing in terms of output size goes aways, if I understood correctly.

Reply:
This is what I meant: https://github.com/JuliaLang/julia/pull/49672#issuecomment-1671160754